client/signup/steps/domains/index.jsx

```diff
@@ -61,6 +61,25 @@ const DomainsStep = React.createClass( {
 		return { products: productsList.get() };
 	},
 
+	componentWillMount: function() {
+		const { stepName, queryObject } = this.props;
+
+		// skip this step if domain name is given in the url via the `domain` query parameter
+		if ( queryObject && queryObject.domain ) {
```

Comment:
I agree it breaks the navigation history and it doesn't look the best. I went for the easy fix first.

> Remove the domains step from this flow entirely and provide the domainItem through the site-or-domain step.

Yep but that might require a bit of change in `site-or-domain` since we use the results from `/products` to create the domainItem (for the product_slug mainly I think).

> This fixes both issues, but we'll need to link the user back to /domains if they want to change their domain selection.

That also might need some sort of a hack. If we imagine porting the `DomainSearch` component from `/domains` to the `domain-first` flow in a near future, it would make sense to just go to the first step which has this component.